Verify in-toto's supply chain with in-toto

SolidifiedRay opened the pull request:
**Fixes issue #**: #278

**Description of the changes being introduced by the pull request**:

Provide a document that includes instructions on how to verify the in-toto's supply chain with the provided in-toto metadata as the user update in-toto.
For more instructions, please refer here: [doc/metadata/layout for sdist/README.md](https://github.com/in-toto/in-toto/pull/444/files#diff-99c924973d4fc6f8553b8d6a26bd118da1fa936d092ca429b42fe2f2e7ef8309)
To bootstrap in-toto, I create a simple example of using TUF to bootstrap in-toto: https://github.com/SolidifiedRay/tuf-bootstrap-in-toto

_*This is no an actual PR. Instead, we may want to discuss the code of the in-toto supply chains layout on this page._

**Current behavior:**
No publicly available in-toto metadata and no instructions to verify in-toto with in-toto.

**Expected behavior:**
Generate metadata and provide instructions how to use it to verify in-toto with in-toto.

SolidifiedRay committed:
Add in-toto supply chains layout

Signed-off-by: SolidifiedRay <yc3346@nyu.edu>

SolidifiedRay committed:
Add doc for verify in-toto's supply chain with in-toto

Signed-off-by: SolidifiedRay <yc3346@nyu.edu>

adityasaky commented on doc/metadata/README.md:
```diff
@@ -0,0 +1,20 @@
+===========================================
+
+**Overview**  
+This document describes how to use in-toto to verify the software supply chains for the PyPi source of in-toto v1.0.1.
```
We should generalize these instructions as much as possible. We can take the tone of "Consider v1.0.1 as an example" or something like that.

adityasaky replied:
Additionally, we may want to either recommend starting from a clean clone or mention some common things to `--exclude`.

adityasaky commented on doc/metadata/README.md:
```diff
@@ -0,0 +1,20 @@
+
+The basic supply chains are laid out in "in_toto_sdist.layout" and consist of two steps:
+
+1. "git-tag" -- Sign the tag using the command "git tag --sign v1.0.1 -m "v1.0.1"".
```
Let's add the in-toto-run wrapped commands here.

adityasaky commented on doc/metadata/README.md:
```diff
@@ -0,0 +1,20 @@
+Verify in-toto's supply chain with in-toto
```
Can we also put some sample links? We can likely sign everything with one key for now and see how verification works.

adityasaky commented on doc/metadata/README.md:
```diff
@@ -0,0 +1,20 @@
+- For the first step, we want to explicitly allow all the files to make sure we tag all the materials files to sign.
+- Since we only sign the tag in the first step, no files will be changed and there will be no products. So we only need to add the "DISALLOW *" rule for the expected products.
+- For the second step, we match all the files with the materials in the first step, because we don't have products in the first step.
+- We may want to use the in-toto-run command to create the links files and be specific for the materials and products, for example, "in-toto-run --step-name create-sdist --materials . --products in-toto.1.0.1.tar.gz --key owner --python setup.py sdist"
```
I wonder if we'll want to `lstrip` the path to the in-toto project on the local machine where this is run.

adityasaky replied:
```suggestion
- We can use the in-toto-run command to create the links files and be specific for the materials and products, for example, "in-toto-run --step-name create-sdist --materials . --products in-toto.1.0.1.tar.gz --key owner -- python setup.py sdist"
```

adityasaky commented on doc/metadata/README.md:
```diff
@@ -0,0 +1,20 @@
+
+- For the bootstrapping/chicken-egg situation, we assume that the user already has an older version of the in-toto locally.
+- For the first step, we want to explicitly allow all the files to make sure we tag all the materials files to sign.
+- Since we only sign the tag in the first step, no files will be changed and there will be no products. So we only need to add the "DISALLOW *" rule for the expected products.
```
```suggestion
- Since we only tag a release in the first step, no files are changed and there will be no products. So we only need to add the "DISALLOW *" rule for the expected products.
```

adityasaky commented on doc/metadata/README.md:
```diff
@@ -0,0 +1,20 @@
+- Since we only sign the tag in the first step, no files will be changed and there will be no products. So we only need to add the "DISALLOW *" rule for the expected products.
+- For the second step, we match all the files with the materials in the first step, because we don't have products in the first step.
+- We may want to use the in-toto-run command to create the links files and be specific for the materials and products, for example, "in-toto-run --step-name create-sdist --materials . --products in-toto.1.0.1.tar.gz --key owner --python setup.py sdist"
+- We may want to add and combine the "Creates binary distribution (.whl)" step with the second step.
```
```suggestion
- TODO: Combine the "Creates binary distribution (.whl)" step with the second step.
```

adityasaky commented on doc/metadata/README.md:
```diff
@@ -0,0 +1,20 @@
+- For the second step, we match all the files with the materials in the first step, because we don't have products in the first step.
+- We may want to use the in-toto-run command to create the links files and be specific for the materials and products, for example, "in-toto-run --step-name create-sdist --materials . --products in-toto.1.0.1.tar.gz --key owner --python setup.py sdist"
+- We may want to add and combine the "Creates binary distribution (.whl)" step with the second step.
+- We may want to attach the layout file to each release.
```
I think this is a broader discussion point. We likely want to make the layout available somewhere other than GitHub?

adityasaky reviewed (commented):
This is looking good so far, @SolidifiedRay, thank you for working on this!

SolidifiedRay committed:
Add sample links

Signed-off-by: SolidifiedRay <yc3346@nyu.edu>

adityasaky commented on doc/metadata/in_toto_sdist.layout:
```diff
@@ -0,0 +1,667 @@
+      "MATERIALS",
+      "FROM",
+      "sign-tag"
+     ]
```
```suggestion
     ],
```

adityasaky commented on doc/metadata/in_toto_sdist.layout:
```diff
@@ -0,0 +1,667 @@
+      "IN",
+      "dist",
+      "FROM",
+      "build-sdist"
```
```suggestion
      "create-sdist"
```

adityasaky commented on doc/metadata/in_toto_sdist.layout:
```diff
@@ -0,0 +1,667 @@
+      "WITH",
+      "MATERIALS",
+      "FROM",
+      "sign-tag"
```
```suggestion
      "git-tag"
```

adityasaky committed:
Multiple changes:
- Rename file to use hyphens
- Fix JSON bug with commas
- Add signed version of layout
- Add keys from in-toto/demo (duplicated in repo)
- Fix step names in layout

Signed-off-by: Aditya Sirish <aditya@saky.in>
Signed-off-by: SolidifiedRay <yc3346@nyu.edu>

SolidifiedRay committed:
Fix typo and re-sign the layout with Alice key

Signed-off-by: SolidifiedRay <yc3346@nyu.edu>

SolidifiedRay committed:
Add in-toto-verify command

Signed-off-by: SolidifiedRay <yc3346@nyu.edu>

adityasaky commented on doc/metadata/README.md:
````diff
@@ -0,0 +1,36 @@
+The example links are laid out in "git-tag.776a00e2.link" and "create-sdist.776a00e2.link". The key for signing the steps is the bob's key from the in-toto demo: https://github.com/in-toto/demo/tree/master/functionary_bob
+
+To verify the supply chain, we can run the in-toto verify command:  
+```in-toto-verify -l in-toto-sdist-signed.layout -k keys/alice.pub -v```
````
The inspection here will fail. We should include the current .tar.gz for the sake of testing, and maybe update that to the source of v1.1.0 when we're ready to release, using that as an example moving forward. The example throughout this guide can be for v1.1.0, I suspect.

adityasaky replied:
@SolidifiedRay and I verified that the inspection works with the source archive present, so essentially we have a working layout for this minimal supply chain!

adityasaky replied:
We did need to update the untar command though, it used a CLI argument to pass in the source archive. We switched to regular `tar` as we do in in-toto/demo.

adityasaky commented on doc/metadata/README.md:
```diff
@@ -0,0 +1,36 @@
+**Todos**
+
+- Combine the "Creates binary distribution (.whl)" step with the second step.
+- We need to decide where to put the layout file. One idea is to attach the layout file to each release, but we probably want to make it available somewhere other than Github.
```
As a point of discussion: what's the best way to handle layout generation? In this instance, Yuanrui started off using the layout web tool and made changes by hand. Do we include that modified version as a template and make sure to keep it updated with changes moving forward? Essentially, we'd have to update artifact rules.

SolidifiedRay committed:
Addthe in-toto tar file and change the inspection command

Signed-off-by: SolidifiedRay <yc3346@nyu.edu>

SolidifiedRay committed:
Fix typo

Signed-off-by: SolidifiedRay <yc3346@nyu.edu>

trishankatdatadog commented:
Good idea! Are there any plans to use TUF to distribute the in-toto supply chain?

SolidifiedRay committed:
Add layout for sdist and bdist

Signed-off-by: SolidifiedRay <yc3346@nyu.edu>

SolidifiedRay committed:
Use unzip instead of untar for the bdist

Signed-off-by: SolidifiedRay <yc3346@nyu.edu>

SolidifiedRay committed:
Add IN rules

Signed-off-by: SolidifiedRay <yc3346@nyu.edu>

adityasaky commented:
I think it's time to revive this PR again. We'd last left off looking at ITE-2 as a means to distribute the metadata.

trishankatdatadog commented:
> I think it's time to revive this PR again. We'd last left off looking at ITE-2 as a means to distribute the metadata.

Would be a good time to dogfood all our stuff!

shibumi commented:
> Good idea! Are there any plans to use TUF to distribute the in-toto supply chain?

@trishankatdatadog I thought about embedding our own TUF root into the rekor instance. According to Dan, rekor is able to handle multiple TUF roots. I spoke with @SantiagoTorres about this and he suggested that we might do a key signing party.

I did some prototyping, you might be able to answer a few of these questions:

![https://kurisu.shibumi.dev/share/in-toto.drawio.png](https://kurisu.shibumi.dev/share/in-toto.drawio.png)

It's just a scratch that has to be discussed, don't take my prototoype for granted :)

trishankatdatadog commented:
Oops, sorry, missed this. WDY need from me?

SolidifiedRay commented:
I created a simple example of using TUF to distribute in-toto metadata: https://github.com/SolidifiedRay/tuf-bootstrap-in-toto

JustinCappos commented:
We should also validate TUF with in-toto metadata....  Ray, would you
kindly create an issue on the TUF tracker?

On Sat, May 14, 2022 at 4:48 AM Ray Chen ***@***.***> wrote:

> I created a simple example of using TUF to distribute in-toto metadata:
> https://github.com/SolidifiedRay/tuf-bootstrap-in-toto
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/in-toto/in-toto/pull/444#issuecomment-1126490684>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAGROD7TFWEAPLL454ZYG7DVJ252FANCNFSM42M2VNJQ>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

adityasaky commented:
https://github.com/theupdateframework/python-tuf/pull/2000 is implementing that courtesy of @lukpueh.

adityasaky commented:
Based on some recent updates: we can automate the release process (as described in the TUF counterpart to this PR) and generate in-toto metadata in the CI workflow. Also, we should be able to use in-toto's github action (with some upcoming updates to improve its UX) to simplify matters. IMO the layout is largely going to look similar in either case (though we can probably simplify some of the rules with succinct patterns).

JustinCappos commented:
Witness already does this! https://github.com/in-toto/witness/actions/runs/8147511831/attempts/1#summary-22268394102
